lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU

```diff
@@ -0,0 +1,252 @@
+	bHasAllStates		: BOOL;
+	bHasAllParameters :BOOL := TRUE;
+	bBusy                : BOOL;
+    bError               : BOOL;
+    nErrId               : UDINT;
+	sErrMsg				 :STRING;
+END_VAR
+VAR_IN_OUT
+	//ARRAY [1.. MOTION_GVL.MAX_STATES] ;
+	arrStates: ARRAY[*] OF ST_DbStateParams;
+	io_fbFFHWO    :    FB_HardwareFFOutput;
+END_VAR
+VAR
+	(* JSON *)
+	fbJson 			: FB_JsonDomParser;
+	jsonProp		: SJsonValue;
+	jsonValue		: SJsonValue;
+	jsonParam		: SJsonValue;
+	jsonnTran      : SJsonValue;
+	jsonnRate      : SJsonValue;
+	Step: INT;
+	index: DINT;
+	nStateCount:DINT;
+    RisingEdge           : R_TRIG;
+
+	(*Logger*)
+	tNewMessage : R_TRIG;
+    fbLogger : FB_LogMessage := (eSubsystem:=E_SubSystem.MPS, nMinTimeViolationAcceptable:=10);	
+	(*FFO*)
+	FFO    :    FB_FastFault :=(
+        i_Desc := 'Fault occurs when there is an error loading safe beam parameters from json file',
+        i_TypeCode := 16#FF13);
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[(* Retrieve JSON content *)
+RisingEdge(CLK:=bExecute); 
+
+CASE Step OF 
+    0:     (* Idle state *) 
+          IF RisingEdge.Q THEN 
+               bBusy := TRUE; 
+               bError:= FALSE; 
+               nErrId:=0; 
+               Step := 20; 
+			   fbLogger(sMsg:='Start Parsing Json DOC.', eSevr:=TcEventSeverity.Info);
+          END_IF 
+	
+	 20: (*Find the Device Name*)  
+		bHasDevice := fbJson.HasMember(jsonDoc, sDeviceName);
+		IF (bHasDevice) THEN
+			jsonProp := fbJson.FindMember(jsonDoc, sDeviceName);
+			IF jsonProp <> 0 THEN	
+				Step := Step+1;
+				ELSE
+					Step := 900 + Step;
+					nErrId := Step; 
+					bError := TRUE; 
+					sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);
+				END_IF
+		ELSE
+			Step := 900 + Step;
+			nErrId := Step; 
+			bError := TRUE; 
+			sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);		
+		END_IF
+		
+	 21: (*Check device States in DOM*)
+	 		FOR Index:=LOWER_BOUND(arrStates,1) TO UPPER_BOUND(arrStates,1) BY 1 DO
+				  IF NOT ((arrStates[Index].sPmpsState= '') OR (arrStates[Index].sPmpsState= 'Invalid')) THEN
+					  IF NOT (fbJson.HasMember(jsonProp, arrStates[Index].sPmpsState)) THEN
+							  Step := 900+ Step;
+							  nErrId := Step; 
+							  bError := TRUE; 
+							  sErrMsg:= CONCAT('Error loading device state in Json File : ',arrStates[Index].sPmpsState);
+							  EXIT;
+					  END_IF
+				   ELSE 
+					   nStateCount :=Index-1;
+					   bHasAllStates :=TRUE;
+					   EXIT;
+				  END_IF
+		    END_FOR
+
+		   Step := Step+1;
+			
+	 22: //MAke sure state array input has State name data
+		  IF (nStateCount =0) THEN
+			  bHasAllStates :=FALSE;
+			  Step := 900+ Step;
+			  nErrId := Step; 
+			  bError := TRUE; 
+			  sErrMsg:= CONCAT('Error Input State array is empty : ',sDeviceName);				  
+		 ELSE 
+			 Step := Step+1;		 
+		 END_IF;
+
+	 23:  (*loop Device State Arrays to find all States*)
+	      FOR Index:=1 TO nStateCount BY 1 DO
```

Comment:
The logic in this section is broken for arrays that don't start at 1 (my current error) and arrays that end at 1 (hypothesis)
I picked array position 0 for the transition states and they aren't loading